streams/src/main/java/org/apache/kafka/streams/state/internals/VersionedStoreQueryUtils.java

```diff
@@ -0,0 +1,189 @@
+                // this topic isn't bounded for our partition, so just skip over it.
+            } else {
+                if (!seenPartitionPositions.containsKey(partition)) {
+                    // we haven't seen a partition that we have a bound for
+                    return false;
+                } else if (seenPartitionPositions.get(partition) < partitionBounds.get(partition)) {
+                    // our current position is behind the bound
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <R> QueryResult<R> runVersionedKeyQuery(final Query<R> query,
+                                                  final PositionBound positionBound,
+                                                  final QueryConfig config,
+                                                  final StateStore store) {
+        if (store instanceof VersionedKeyValueStore) {
+            final VersionedKeyValueStore<Bytes, byte[]> versionedKeyValueStore =
+                (VersionedKeyValueStore<Bytes, byte[]>) store;
+            if (query instanceof VersionedKeyQuery) {
+                final VersionedKeyQuery<Bytes, byte[]> rawKeyQuery =
+                    (VersionedKeyQuery<Bytes, byte[]>) query;
+                try {
+                    final VersionedRecord<byte[]> bytes;
+                    if (((VersionedKeyQuery<?, ?>) query).asOfTimestamp().isPresent()) {
+                        bytes = versionedKeyValueStore.get(rawKeyQuery.key(),
+                            ((VersionedKeyQuery<?, ?>) query).asOfTimestamp().get()
+                                .getLong(ChronoField.INSTANT_SECONDS));
+                    } else {
+                        bytes = versionedKeyValueStore.get(rawKeyQuery.key());
+                    }
+                    return (QueryResult<R>) QueryResult.forResult(bytes);
+                } catch (final Exception e) {
+                    final String message = parseStoreException(e, store, query);
+                    return QueryResult.forFailure(
+                        FailureReason.STORE_EXCEPTION,
+                        message
+                    );
+                }
+            } else {
+                // Here is preserved for VersionedMultiTimestampedKeyQuery
+                return QueryResult.forUnknownQueryType(query, store);
+            }
+        } else {
+            return QueryResult.forUnknownQueryType(query, store);
+        }
+    }
+
+    public static <V> VersionedRecord<V> deserializeVersionedRecord(final StateSerdes<?, V> serdes, final VersionedRecord<byte[]> rawVersionedRecord) {
+
+        final Deserializer<V> valueDeserializer = serdes.valueDeserializer();
+        final long timestamp = rawVersionedRecord.timestamp();
+        final V value = valueDeserializer.deserialize(serdes.topic(), rawVersionedRecord.value());
+        valueDeserializer.close();
```

Comment:
`valueDesarializer()` does not create a new instance/object, but returns a shared object -- we should not close it.